File: kruskal/dark-roads.cpp
#include <cstdio>
#include <cstring>
#include <cstdlib>

using namespace std;

#define scanner2(m, n) scanf("%d%d", &m, &n)
#define scanner3(x, y, z) scanf("%d%d%d", &x, &y, &z)
struct Edge { int x, y, z; };

int amount;
int vertices[999999];
Edge edges[999999];

// Creating individual sets
void makeSet(int n) {
    for (int i = 0; i < n; i++)
        vertices[i] = i;
}

// Getting the sets created
int findSet(int n) {
    if(vertices[n] != n)
        vertices[n] = findSet(vertices[n]);
    return vertices[n];
}

// Union of sets
int unionSet(int x, int y, int p) {
    int i = findSet(x), j = findSet(y);

    if(i != j) {
        amount -= edges[p].z;
        if(i > j) vertices[i] = j;
        else vertices[j] = i;
        return 1;
    }

    return 0;
}

// (a,b) => a.z - b.z
int compare(const void *a, const void *b) {
    return (*(struct Edge *)a).z - (*(struct Edge *)b).z;
}

int main(int argc, char const *argv[]) {
    int m, n, i;

    while(scanner2(m, n) && (m || n)) {
        amount = 0;
        memset(&edges, 0, sizeof(edges));

        for (i = 0; i < n; i++) {
            scanner3(edges[i].x, edges[i].y, edges[i].z);
            amount += edges[i].z;
        }

        makeSet(m);
        qsort(edges, n, sizeof(edges[0]), compare);
        for (i = 0; i < n; ++i)
            unionSet(edges[i].x, edges[i].y, i);
        printf("%d\n", amount);
    }

    return 0;
}

File: kruskal/dark-roads.py
m, n = map(int, raw_input().split())
amount = 0
edges = []
vertices = []


def find_set(n):
    if vertices[n] != n:
        vertices[n] = find_set(vertices[n])
    return vertices[n]


def union_set(x, y, p):
    vertice1 = find_set(x)
    vertice2 = find_set(y)
    if vertice1 != vertice2:
        global amount
        amount -= edges[p][0]
        if vertice1 > vertice2:
            vertices[vertice1] = vertice2
        else:
            vertices[vertice2] = vertice1

while(True):
    params = map(int, raw_input().split())
    if params[0] == 0 and params[1] == 0: 
        break
    else:
        x = params[0]
        y = params[1]
        z = params[2]
        edges.append((z, x, y))
        amount += z

vertices = range(m+1)

edges.sort()

for i in xrange(n):
    union_set(edges[i][1], edges[i][2], i)

print(amount)
